framework/src/controller/bus.ts

```diff
@@ -0,0 +1,367 @@
+	public events: { [key: string]: boolean };
+	public channels: {
+		[key: string]: ChannelInfo;
+	};
+	public rpcClients: { [key: string]: ReqSocket };
+	public pubSocket?: PubEmitterSocket;
+	public subSocket?: SubEmitterSocket;
+	public rpcSocket?: RepSocket;
+	public rpcServer?: RPCServer;
+	public channel?: BaseChannel;
+
+	public constructor(
+		options: object,
+		logger: Logger,
+		config: BusConfiguration,
+	) {
+		super(options);
+		this.logger = logger;
+		this.config = config;
+
+		// Hash map used instead of arrays for performance.
+		this.actions = {};
+		this.events = {};
+		this.channels = {};
+		this.rpcClients = {};
+	}
+
+	public async setup(): Promise<boolean> {
+		if (!this.config.ipc.enabled) {
+			return true;
+		}
+
+		this.pubSocket = axon.socket('pub-emitter') as PubEmitterSocket;
+		this.pubSocket.bind(this.config.socketsPath.pub);
+
+		this.subSocket = axon.socket('sub-emitter') as SubEmitterSocket;
+		this.subSocket.bind(this.config.socketsPath.sub);
+
+		this.rpcSocket = axon.socket('rep') as RepSocket;
+		this.rpcServer = new RPCServer(this.rpcSocket);
+		this.rpcSocket.bind(this.config.socketsPath.rpc);
+
+		this.rpcServer.expose(
+			'registerChannel',
+			(moduleAlias, events, actions, options, cb) => {
+				this.registerChannel(moduleAlias, events, actions, options)
+					.then(() => cb(null))
+					.catch(error => cb(error));
+			},
+		);
+
+		this.rpcServer.expose('invoke', (action, cb) => {
+			this.invoke(action)
+				.then(data => cb(null, data))
+				.catch(error => cb(error));
+		});
+
+		this.rpcServer.expose('invokePublic', (action, cb) => {
+			this.invokePublic(action)
+				.then(data => cb(null, data))
+				.catch(error => cb(error));
+		});
+
+		await Promise.race([
+			this._resolveWhenAllSocketsBound(),
+			this._rejectWhenAnySocketFailsToBind(),
+			this._rejectWhenTimeout(SOCKET_TIMEOUT_TIME),
+		]);
+		await this._removeAllListeners();
+
+		return true;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/require-await
+	public async registerChannel(
+		moduleAlias: string,
+		// Events should also include the module alias
+		events: EventsArray,
+		actions: ActionsObject,
+		options: RegisterChannelOptions,
+	): Promise<void> {
+		events.forEach(eventName => {
+			if (this.events[`${moduleAlias}:${eventName}`] !== undefined) {
+				throw new Error(`Event "${eventName}" already registered with bus.`);
+			}
+			this.events[`${moduleAlias}:${eventName}`] = true;
+		});
+
+		Object.keys(actions).forEach(actionName => {
+			if (this.actions[`${moduleAlias}:${actionName}`] !== undefined) {
+				throw new Error(`Action "${actionName}" already registered with bus.`);
+			}
+
+			this.actions[`${moduleAlias}:${actionName}`] = actions[actionName];
+		});
+
+		let { channel } = options;
+
+		if (options.rpcSocketPath) {
+			const rpcSocket = axon.socket('req') as ReqSocket;
+			rpcSocket.connect(options.rpcSocketPath);
+
+			// TODO: Fix this override
+			// eslint-disable-next-line
+			// @ts-ignore
```

Comment:
why this needs `ts-ignore`?

Reply:
Its completely wrong logic down there... 

```
			channel = new RPCClient(rpcSocket);
```

overriding the channel object passed from options as type `BaseChannel`. Already posted a comment to fix it.